### PR TITLE
improve MPI-IO support

### PR DESCRIPTION
**Pull request by MichaelBrim**
### Description

This PR improves our MPI-IO support by using the latest version of Gotcha, which seems to properly deal with dlopen'ed libraries. SpectrumMPI/OpenMPI dlopens its MPI-IO library on the first call to an MPI-IO API method. Also needed were wrappers for `statfs()` and `fstatfs()`, which the ROMIO library uses to detect the target filesystem and choose an appropriate implementation. The PR also includes improvements for MPI-IO support in the examples.

A brief summary of the changes:
* added wrappers for statfs() and fstatfs()
* migrated gotcha support to latest API
* travis update for new gotcha@1.0.3 spack package
* fixed examples to not break when Unify configured with `--enable-mpi-mount`. previously, they ended up calling `unifyfs_mount()` twice and the second call failed.
* added MPI-IO support to testutil, so all of the newer examples based on testutil can seamlessly use MPI-IO as the I/O mechanism

### Motivation and Context

This enables UnifyFS to work with MPI-IO on SpectrumMPI systems like Summit/Sierra when using Gotcha for wrapping. The linker-based wrapping still does not work to intercept MPI-IO on these systems.

### How Has This Been Tested?

Tested on summitdev with a few external MPI-IO programs and the updated `writeread-gotcha` example using MPI-IO mode (`--mpiio`). When using external programs, it is important that `libunifyfs_gotcha.so` gets linked before the MPI libraries.

### Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Performance enhancement (non-breaking change which improves efficiency)
- [x] Code cleanup (non-breaking change which makes code smaller or more readable)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [x] Testing (addition of new tests or update to current tests)
- [ ] Documentation (a change to man pages or other documentation)

### Checklist:

- [x] My code follows the UnifyFS code style requirements.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.
- [x] All commit messages are properly formatted.

### CODE REVIEW

File: client/src/unifyfs-sysio.c [hunk 1]
```diff
@@ -527,6 +527,76 @@ int UNIFYFS_WRAP(__fxstat)(int vers, int fd, struct stat* buf)
 }
 #endif
 
+
+#ifdef HAVE_SYS_STATFS_H
+
+/* tmpfs seems like a safe choice for something like UnifyFS */
+#ifndef TMPFS_MAGIC
+#define TMPFS_MAGIC 0x01021994
+#endif
```
Comment (tonyhutter) [inline client/src/unifyfs-sysio.c:536]: I think we're cool enough to have our own f_type :sunglasses:      Maybe?:
```
0x554E4946  
   U N I F
```

**Comment by adammoody** [comment]
This is great, @MichaelBrim !  Nice work!

**Comment by MichaelBrim** [comment]
needs an update for the recent unifyfs_intercept_path() API change, and a few checkpatch errors

File: client/src/unifyfs-sysio.c (hunk 1 above)
Comment (MichaelBrim) [reply client/src/unifyfs-sysio.c:536]: I'll leave that decision up to the people with the money.

**Comment by adammoody** [comment]
@CamStan , would you mind updating the bootstrap to the proper gotcha version when you update the docs?

**Comment by CamStan** [comment]
Will do! Got a list of updates for the docs going that I'll throw in a PR.

**Pull request merged**